### PR TITLE
adds initial git commit iteration script

### PR DESCRIPTION
**Pull request by jregchiu**
Git commit iteration script implementation:
- based off the current HEAD of the specified repo
- creates a new temp working tree/directory and branch FOR SAFETY!
- iterates over each commit from the branch, excluding merge commits
- cats the commit metadata (hash/author/date) to the specified output file
- cleans up working tree/directory and branch after

### CODE REVIEW

File: git-iterate [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
+
+# Iterate over all the commits in the current branch of a repo
+# This script should be placed and run from outside the repo
+# Usage: ./git-iterate <repo-path> <temp-work-dir> <output-file>
```
Comment (lydiachoi) [inline git-iterate]: can you throw a line in here of an actual command for an example

File: git-iterate [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+#
+# repo-path:     relative path to dir of repo
+# temp-work-dir: temporary work dir/tree to use
```
Comment (lydiachoi) [inline git-iterate]: still a little unclear what this means - different than all branch names where?

File: git-iterate [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#                MUST BE DIFFERENT THAN ALL CURRENT BRANCH NAMES
+# output-file:   output file to redirect results to
```
Comment (lydiachoi) [inline git-iterate:8]: is this the file name or file path?

File: git-iterate [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+if [[ $# -ne 3 ]]; then
+    echo "usage: ./git-iterate <repo-path> <temp-work-dir> <output-file>"
+    exit 2
```
Comment (lydiachoi) [inline git-iterate:16]: ELI5: what does 2 do?

File: git-iterate [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+fi
+
+# create a new work dir/tree based on the repo's HEAD, in the same dir as the script
+# this also creates and checks out a new branch with the same name as the work dir/tree
+# this stuff is FOR SAFETY!!!
+git --git-dir=$1/.git worktree add $2
```
Comment (lydiachoi) [inline git-iterate]: ELi5; still confused, what does this do? we're checking out a new branch so that we're not messing with the existing one - is there an action we have to do to merge after?

File: git-iterate [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    # NOTE: these scripts might need arg $2 as well, since that's the work dir with the files
+    # ./memory.sh
+    # ./teststuff.sh >> $3
+
+    # clean up the files back to the original commit state before switching
+    git -C $2 checkout -q .
+done
+
+# clean up the temporary work dir/tree and branch
+git -C $1 worktree remove $2
```
Comment (lydiachoi) [inline git-iterate]: clean up by removing? does the work get saved in the text file and then we delete the branch? If we delete the branch can that branch name be reused?

**Review by lydiachoi — APPROVED** [review]
few comments,

File: git-iterate (hunk 1 above)
Comment (jregchiu) [reply git-iterate]: Added an example with actual usage and explanation

File: git-iterate (hunk 2 above)
Comment (jregchiu) [reply git-iterate]: Simplified by removing the need for the temp-work-dir as a parameter, since it's fully internal to the script implementation.

File: git-iterate (hunk 3 above)
Comment (jregchiu) [reply git-iterate:8]: Relative file path. Made clearer with the actual example usage.

File: git-iterate (hunk 4 above)
Comment (jregchiu) [reply git-iterate:16]: That's just convention - should exit with a non-0 status to indicate an error.

File: git-iterate (hunk 5 above)
Comment (jregchiu) [reply git-iterate]: We make a copy of the working directory (the files in the repo themselves) and a new branch, so that it's totally separate from the original working directory and branches in the repo. This way any changes we make for the testing stuff will not affect the original working directory and branches in the repo.

We just delete the working directory and branch after because there's no changes we want to commit and merge.

File: git-iterate (hunk 6 above)
Comment (jregchiu) [reply git-iterate]: Yes, we clean up by deleting the temporary work directory and branch. Everything is redirected into the output file which is in the same directory as the script itself.
